File: go/jzoffer/lengthOfLongestSubstring_48.go
package jzoffer

// 字符不一定是 a -z
func lengthOfLongestSubstring(s string) int {
	preIndexes := make(map[int32]int)

	curLen, maxLen := 0, 0
	for i, char := range s {
		if preIndex, ok := preIndexes[char-'a']; !ok || i-preIndex > curLen {
			curLen++
		} else {
			curLen = i - preIndexes[char-'a']
		}

		preIndexes[char-'a'] = i
		maxLen = max(maxLen, curLen)
	}
	return maxLen
}

File: java/jzoffer/LengthOfLongestSubstring_48.java
import java.util.Arrays;

public class LengthOfLongestSubstring_48 {

    public int lengthOfLongestSubstring(String s) {
        int[] preIndexes = new int[255];
        Arrays.fill(preIndexes, -1);

        int res = 0, len = 0;

        for (int i = 0; i < s.length(); i++) {
            // 在当前距离之外 或者 没有访问过
            if (preIndexes[s.charAt(i)] == -1 || i - preIndexes[s.charAt(i)] > len) {
                len++;
            } else {
                // 在当前距离之类
                len = i - preIndexes[s.charAt(i)];
            }
            res = Math.max(res, len);
            preIndexes[s.charAt(i)] = i;
        }
        return res;
    }

    public static void main(String[] args) {
        System.out.println(new LengthOfLongestSubstring_48().lengthOfLongestSubstring("pwwkew"));
    }
}
